lib/puma/launcher.rb

```diff
@@ -42,26 +42,39 @@ class Launcher
     #   end
     #   Puma::Launcher.new(conf, log_writer: Puma::LogWriter.stdio).run
     def initialize(conf, launcher_args={})
-      @runner        = nil
-      @log_writer    = launcher_args[:log_writer] || LogWriter::DEFAULT
-      @events        = launcher_args[:events] || Events.new
-      @argv          = launcher_args[:argv] || []
-      @original_argv = @argv.dup
-      @config        = conf
-
-      env = launcher_args.delete(:env) || ENV
+      ## Minimal initialization for a potential early restart (e.g. when pruning bundle)
 
+      @config = conf
       @config.clamp
+
       @options = @config.options
 
+      @log_writer = launcher_args[:log_writer] || LogWriter::DEFAULT
+      @log_writer.formatter = LogWriter::PidFormatter.new if clustered?
+      @log_writer.formatter = @options[:log_formatter] if @options[:log_formatter]
+      @log_writer.custom_logger = @options[:custom_logger] if @options[:custom_logger]
       @options[:log_writer] = @log_writer
       @options[:logger] = @log_writer if clustered?
 
+      @events = launcher_args[:events] || Events.new
+
+      @argv = launcher_args[:argv] || []
+      @original_argv = @argv.dup
+
+      ## End minimal initialization
```

Comment:
Request to say less or say more. Either explain what this is without someone needing to reference the history, preferably with the context of "how to know if you're putting your config in the right place or not" in mind or remove it:

```suggestion
```

Reply:
I think the key is that `log_config` must go after anything that would spawn a new process like `prune_bundler!`. Maybe we should put that comment on `prune_bundler!` or `log_config` (or both).